packages/transactions/src/builders.ts

```diff
@@ -1692,3 +1723,70 @@ export async function estimateTransactionFeeWithFallback(
     throw error;
   }
 }
+
+/** @internal multi-sig signing re-use */
+function mutatingSignAppendMultiSig(
+  /** **Warning:** method mutates `transaction` */
+  transaction: StacksTransaction,
+  publicKeys: string[],
+  signerKeys: string[],
+  address?: string
+) {
+  if (isSingleSig(transaction.auth.spendingCondition)) {
+    throw new Error('Transaction is not a multi-sig transaction');
+  }
+
+  const signer = new TransactionSigner(transaction);
+
+  const pubs = ensureValidMultiSigPublicKeyOrder(
+    publicKeys,
+    transaction.auth.spendingCondition.signaturesRequired,
+    transaction.auth.spendingCondition.hashMode,
+    address ? createAddress(address).hash160 : undefined
+  );
+
+  // sign in order of public keys
+  for (const publicKey of pubs) {
+    const signerKey = signerKeys.find(key => bytesToHex(pubKeyfromPrivKey(key).data) === publicKey);
+    if (signerKey) {
+      // either sign and append message signature (which allows for recovering the public key)
+      signer.signOrigin(createStacksPrivateKey(signerKey));
+    } else {
+      // or append the public key (which did not sign here)
+      signer.appendOrigin(publicKeyFromBytes(hexToBytes(publicKey)));
+    }
+  }
+}
+
+/** @internal Get the matching public-keys array for a multi-sig address */
+function ensureValidMultiSigPublicKeyOrder(
+  publicKeys: string[],
+  numSigs: number,
+  hashMode: MultiSigHashMode,
+  hash?: string
```

Comment:
There's really no point in calling this function without the address

Reply:
That's true, this was to push the check down, but I can try to improve this 🤔